### PR TITLE
SCP-2755: Chain Index UTXO state DB backend

### PR DESCRIPTION
**Pull request by sjoerdvisscher**
Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-chain-index/src/Plutus/ChainIndex/UtxoState.hs [hunk 1]
```diff
@@ -159,24 +169,23 @@ data RollbackResult a =
...
-viewTip = tip . measure
-
 -- | Perform a rollback on the utxo index
 rollback :: Point
          -> UtxoIndex TxUtxoBalance
          -> Either RollbackFailed (RollbackResult TxUtxoBalance)
+rollback PointAtGenesis (viewTip -> TipAtGenesis) = Right (RollbackResult TipAtGenesis mempty)
 rollback _ (viewTip -> TipAtGenesis) = Left RollbackNoTip
 rollback targetPoint idx@(viewTip -> currentTip)
+    -- Already at the target point
```
Comment (sjoerdvisscher) [inline plutus-chain-index/src/Plutus/ChainIndex/UtxoState.hs:179]: When resuming from a point, the cardano chain sync service always sends a rollback command from that point, even if the point is the current tip. From that it's clear that a rollback to the current tip should be allowed. Should this also be done in the rollback function in TxIdState? @j-mueller @silky

Comment (silky) [reply plutus-chain-index/src/Plutus/ChainIndex/UtxoState.hs:179]: Yes I believe so. (Ideally we wouldn't even have two separate rollback functions, that could be refactored if inspiration strikes you.)

File: plutus-chain-index/app/Main.hs [hunk 2]
```diff
@@ -107,15 +112,28 @@ chainSyncHandler trace mState conn
       Left err    ->
         logError trace (ConversionFailed err)
       Right txs ->
-        runChainIndex trace mState conn $ appendBlock (tipFromCardanoBlock block) txs
+        void $ runChainIndex trace mState conn $ appendBlock (tipFromCardanoBlock block) txs
 chainSyncHandler trace mState conn
   (RollBackward point _) _ = do
+    putStr "Rolling back to "
```
Comment (j-mueller) [inline plutus-chain-index/app/Main.hs:118]: Have you ever seen this message printed?

File: plutus-chain-index/src/Plutus/ChainIndex/Emulator/DiskState.hs [hunk 3]
```diff
@@ -99,4 +99,6 @@ diagnostics DiskState{_DataMap, _ScriptMap, _TxMap, _RedeemerMap, _AddressMap} =
         , numScripts = toInteger $ Map.size _ScriptMap
         , numAddresses = toInteger $ Map.size $ _unCredentialMap _AddressMap
         , someTransactions = take 10 $ fmap fst $ Map.toList _TxMap
+        , numUnmatchedInputs = 0
```
Comment (j-mueller) [inline plutus-chain-index/src/Plutus/ChainIndex/Emulator/DiskState.hs:102]: Can we get the numbers from the `UtxoIndex TxUtxoBalance`?

Comment (j-mueller) [reply plutus-chain-index/src/Plutus/ChainIndex/Emulator/DiskState.hs:102]: Never mind, it gets added in `Handlers.hs` below - worth a comment

File: plutus-chain-index/src/Plutus/ChainIndex/UtxoState.hs [hunk 4]
```diff
@@ -159,31 +170,58 @@ data RollbackResult a =
...
-
 -- | Perform a rollback on the utxo index
 rollback :: Point
          -> UtxoIndex TxUtxoBalance
          -> Either RollbackFailed (RollbackResult TxUtxoBalance)
-rollback _ (viewTip -> TipAtGenesis) = Left RollbackNoTip
-rollback targetPoint idx@(viewTip -> currentTip)
+rollback = rollbackWith const
+
+rollbackWith
```
Comment (j-mueller) [inline plutus-chain-index/src/Plutus/ChainIndex/UtxoState.hs:179]: haddocks

File: plutus-chain-index/src/Plutus/ChainIndex/UtxoState.hs [hunk 5]
```diff
@@ -159,31 +170,58 @@ data RollbackResult a =
...
+-- | Reduce the number of 'UtxoState's. The given number is the minimum, the index is reduced when it larger than twice that size.
+-- The new index is prefixed with one 'UtxoState' that contains the combined state of the removed 'UtxoState's.
+reduceBlockCount :: Monoid a => Int -> UtxoIndex a -> ReduceBlockCountResult a
+reduceBlockCount minCount ix
+    | utxoBlockCount ix <= 2 * minCount = BlockCountNotReduced
+    | otherwise =
+        let (old, keep) = FT.split ((> (utxoBlockCount ix - minCount)) . getSum . fst) ix
+            combinedState = utxoState old
+        in ReduceBlockCountResult
+            { reducedIndex = combinedState FT.<| keep
```
Comment (j-mueller) [inline plutus-chain-index/src/Plutus/ChainIndex/UtxoState.hs:225]: 👍

File: plutus-chain-index/src/Plutus/ChainIndex/UtxoState.hs [hunk 6]
```diff
@@ -96,14 +103,15 @@ fromTx tx =
         , _tubUnmatchedSpentInputs = Set.mapMonotonic txInRef (view citxInputs tx)
         }
 
-type UtxoIndex a = FingerTree (UtxoState a) (UtxoState a)
-instance Monoid a => Measured (UtxoState a) (UtxoState a) where
-    measure = id
+type UtxoIndex a = FingerTree (Sum Int, UtxoState a) (UtxoState a)
```
Comment (j-mueller) [inline plutus-chain-index/src/Plutus/ChainIndex/UtxoState.hs:106]: We could move the block count to `UtxoState` and give it a name

Comment (j-mueller) [reply plutus-chain-index/src/Plutus/ChainIndex/UtxoState.hs:106]: Or define `newtype BlockCount = BlockCount Int deriving (Semigroup, Monoid) via (Sum Int)`  or something

File: plutus-chain-index/app/Main.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjoerdvisscher) [reply plutus-chain-index/app/Main.hs:118]: Yes, you get one when resuming, but otherwise no. I guess you only see one when it happens live?

**Review by j-mueller — APPROVED** [review]
LGTM!

**Pull request merged**